syntax = "proto3";

service DatabaseService {
  // Створення таблиці
  rpc CreateTable (CreateTableRequest) returns (CreateTableResponse);

  // Видалення таблиці
  rpc DeleteTable (DeleteTableRequest) returns (DeleteTableResponse);

  // Створення рядка
  rpc CreateRow (CreateRowRequest) returns (CreateRowResponse);

  // Редагування рядка
  rpc EditRow (EditRowRequest) returns (EditRowResponse);

  // Отримати всі рядки
  rpc GetAllRows (GetAllRowsRequest) returns (GetAllRowsResponse);

  // Порівняння таблиць
  rpc TableDifference (TableDifferenceRequest) returns (TableDifferenceResponse);

  // Збереження та завантаження бази даних
  rpc SaveDatabase (SaveDatabaseRequest) returns (SaveDatabaseResponse);
  rpc LoadDatabase (LoadDatabaseRequest) returns (LoadDatabaseResponse);
}

// Повідомлення для створення таблиці
message CreateTableRequest {
  string table_name = 1;
  repeated Field fields = 2;
}

message CreateTableResponse {
  string message = 1;
}

message DeleteTableRequest {
  string table_name = 1;
}

message DeleteTableResponse {
  string message = 1;
}

// Повідомлення для створення рядка
message CreateRowRequest {
  string table_name = 1;
  repeated Value values = 2;
}

message CreateRowResponse {
  string message = 1;
}

message EditRowRequest {
  string table_name = 1;
  int32 row_id = 2;
  repeated Value values = 3;
}

message EditRowResponse {
  string message = 1;
}

// Повідомлення для порівняння таблиць
message TableDifferenceRequest {
  string table1_name = 1;
  string table2_name = 2;
}

message TableDifferenceResponse {
  repeated Row difference = 1;
}

// Повідомлення для збереження/завантаження бази даних
message SaveDatabaseRequest {}

message SaveDatabaseResponse {
  string message = 1;
}

message LoadDatabaseRequest {}

message LoadDatabaseResponse {
  string message = 1;
}

// Структури даних
message Field {
  string name = 1;
  string type = 2;  // 'integer', 'real', 'char', 'string', 'date', 'dateInvl'
}

message Value {
  string field_name = 1;
  string value = 2;  // Можна зберігати значення у вигляді рядка і потім конвертувати
}

message Row {
  int32 id = 1;
  repeated Value values = 2;
}

message GetAllRowsRequest {
  string table_name = 1;
}

message GetAllRowsResponse {
  repeated Row rows = 1;
}
